cirq/linalg/decompositions.py

```diff
@@ -363,20 +363,25 @@ def interaction_matrix(m: np.ndarray, c: float) -> np.ndarray:
             before)
 
 
-def kak_canonicalize_vector(x: float, y: float, z: float) -> KakDecomposition:
+def kak_canonicalize_vector(x: float, y: float, z: float,
+                            atol: float = 1e-8) -> KakDecomposition:
     """Canonicalizes an XX/YY/ZZ interaction by swap/negate/shift-ing axes.
 
     Args:
         x: The strength of the XX interaction.
         y: The strength of the YY interaction.
         z: The strength of the ZZ interaction.
+        atol: If any of the coefficients are less than atol, they will be
+        rounded down to zero
 
     Returns:
         The canonicalized decomposition, with vector coefficients (x2, y2, z2)
         satisfying:
 
-            0 ≤ abs(z2) ≤ y2 ≤ x2 ≤ π/4
-            z2 ≠ -π/4
+            0 ≤ z2 ≤ y2 ≤ x2 < π/2
+            x2 + y2 ≤ π/2
+            if z2 =0, kx ≤ π/4
```

Comment:
I don't like this change to the canonical region.

1. It behaves surprisingly under small perturbations. For example, `XX**0.5 @ YY**0.4 @ ZZ**-epsilon` produces `(0.63, 0.63, tiny)` but then `XX**0.5 @ YY**0.4 @ ZZ**+epsilon` causes the X coefficient to jump up to `0.94`. Perturbing smallest interaction by a tiny amount caused the largest interaction to jump by 50%. Previously the only instability of this type was in the sign of the smallest interaction, and it only happened at the X=max plane. AFAIK you never saw magnitudes jumping around (bugs notwithstanding).

2. We have code in the code base that assumes "small interaction coefficient" is equivalent to "small two-qubit operation", and this code will break in the presence of values like pi/2 which actually correspond to small two-qubit operations.

3. The shape of the canonical region is no longer symmetric (with the exception of the X=max plane) about a plane passing through the origin.

    Current shape:

    ![image](https://user-images.githubusercontent.com/79941/58597959-2bb70800-822f-11e9-9a15-b55702e38d21.png)

    Proposed shape:

    ![image](https://user-images.githubusercontent.com/79941/58597991-4db08a80-822f-11e9-8423-baefa94066b6.png)

    The complications of the new shape are particularly apparent in the fact that you needed an additional inequality in order to specify it.

Is this change to the region actually required? I would strongly prefer to stay with the existing region (except for the X=max plane where we were failing to canonicalize correctly, of course).

Reply:
The only reason I switched to this region is because the reference mentioned that there is a one-to-one correspondence between points in this set and equivalence classes in SU(4). It may also be possible to refactor the current region to maintain your "smallness" assumption and still achieve the one-to-one correspondence.

Reply:
I agree that we want a one-to-one correspondence. But we can massage the shape of the region to be better.

Reply:
I've got a variation that changes our current region minimally. The only change is that if the XX coefficient = pi/4 within some tolerance, the ZZ term is forced to be positive. This change also makes @babbush's tests pass. I also verified that forcing the ZZ term to be negative works but arbitrarily chose positive since we have the freedom to choose.